### PR TITLE
Added function to allow users who are assigned to allocations to also be added to the allocation's project

### LINKED ISSUE #16: Add User to Group adds the User only to the Allocation, should add to Project too
Access to the resource allocation will work fine, but the user won't see it show up in their ColdFront list.

### PR DESCRIPTION
**Pull request by QuanMPhm**
Implemented a function in scim_v2/groups.py which adds the user both to the allocation and group. Group users are required to have a status (i.e Active, Removed), role (PI, User, etc), and a choice to enable email notifications. The role and notification values are set to a default of User and False.

Closes #16

### CODE REVIEW

File: src/coldfront_plugin_api/scim_v2/groups.py [hunk 1]
```diff
@@ -59,13 +60,20 @@ def get(self, request, pk, format=None):
         return Response(allocation_to_group_view(allocation))
 
     def patch(self, request, pk, format=None):
+
+        print("Got Data")
```
Comment (knikolla) [inline src/coldfront_plugin_api/scim_v2/groups.py:64]: You have forgotten to remove this.

File: src/coldfront_plugin_api/scim_v2/groups.py [hunk 2]
```diff
@@ -59,13 +60,20 @@ def get(self, request, pk, format=None):
...
+
         if (
                 request.data["schemas"] != ["urn:ietf:params:scim:api:messages:2.0:PatchOp"]
                 or request.data.get("path", "members") != "members"
         ):
             return Response(status=400)
+        
+        print("Processing")
```
Comment (knikolla) [inline src/coldfront_plugin_api/scim_v2/groups.py:72]: You have forgotten to remove this.

File: src/coldfront_plugin_api/scim_v2/groups.py [hunk 3]
```diff
@@ -95,6 +108,10 @@ def patch(self, request, pk, format=None):
                     signals.allocation_remove_user.send(
                         sender=self.__class__, allocation_user_pk=au.pk,
                     )
+
+                    self._set_user_on_project(
+                        project, user, "Removed", "User", False
```
Comment (knikolla) [inline src/coldfront_plugin_api/scim_v2/groups.py:113]: Do not remove a user from the project. They might be in other resource allocations of the project.

File: src/coldfront_plugin_api/scim_v2/groups.py [hunk 4]
```diff
@@ -117,3 +134,26 @@ def _set_user_status_on_allocation(allocation, user, status):
                 status=AllocationUserStatusChoice.objects.get(name=status)
             )
         return au
+    
+    @staticmethod
+    def _set_user_on_project(project, user, status, role, enable_notifications):
+        pu = ProjectUser.objects.filter(
+            project=project,
+            user=user
+        ).first() # Why first, can't we assume this is always unique?
```
Comment (knikolla) [inline src/coldfront_plugin_api/scim_v2/groups.py:143]: That's a good question, but unless you are answering it here, remove the comment.

File: src/coldfront_plugin_api/scim_v2/users.py [hunk 5]
```diff
@@ -38,7 +38,7 @@ class ListUsers(APIView):
     * Only admin users are able to access this view.
     """
     authentication_classes = auth.AUTHENTICATION_CLASSES
-    permission_classes = [IsAdminUser]
+    permission_classes = [IsAdminUser]
```
Comment (knikolla) [inline src/coldfront_plugin_api/scim_v2/users.py:41]: You have added some whitespace here by mistake, your Pull Requests shouldn't introduce unrelated changes like this.

File: src/coldfront_plugin_api/urls.py [hunk 6]
```diff
@@ -8,7 +8,6 @@
 from coldfront_plugin_api import auth, serializers
 from coldfront_plugin_api.scim_v2 import groups, users
 
-
```
Comment (knikolla) [inline src/coldfront_plugin_api/urls.py:11]: You have remove whitespace here by mistake, your Pull Requests shouldn't introduce unrelated changes like this.

**Review by knikolla — CHANGES_REQUESTED** [review]
Looks good, except some minor comments for polish.

Please use descriptive titles with regards to what the PR is solving, rather than addressing another issue by number, which requires the reviewer to search for the issue number.

**Comment by QuanMPhm** [comment]
@knikolla I've made changed to my PR, per your feedback

File: src/coldfront_plugin_api/scim_v2/groups.py [hunk 7]
```diff
@@ -86,6 +89,11 @@ def patch(self, request, pk, format=None):
                     signals.allocation_activate_user.send(
                         sender=self.__class__, allocation_user_pk=au.pk,
                     )
+
+                    self._set_user_on_project(
```
Comment (knikolla) [inline src/coldfront_plugin_api/scim_v2/groups.py:93]: Thinking a bit more about this. I would place this above `au = self._set_user_status_on_allocation` as that is the ordering of operations that ColdFront prefers.

File: src/coldfront_plugin_api/scim_v2/groups.py [hunk 8]
```diff
@@ -86,6 +89,11 @@ def patch(self, request, pk, format=None):
...
+                        project, user, "Active", "User", False
```
Comment (knikolla) [inline src/coldfront_plugin_api/scim_v2/groups.py:94]: Does `enable_notifications` default to false when adding someone from the web dashboard? Where is this default coming from?

File: src/coldfront_plugin_api/scim_v2/groups.py [hunk 9]
```diff
@@ -117,3 +125,26 @@ def _set_user_status_on_allocation(allocation, user, status):
...
+    def _set_user_on_project(project, user, status, role, enable_notifications):
+        pu = ProjectUser.objects.filter(
+            project=project,
+            user=user
+        ).first()
+
+        if pu:
+            pu.status = ProjectUserStatusChoice.objects.get(name=status)
+            pu.role = ProjectUserRoleChoice.objects.get(name=role)
+            pu.enable_notifications = enable_notifications
```
Comment (knikolla) [inline src/coldfront_plugin_api/scim_v2/groups.py:139]: This line would override the `enable_notifications` value that the ProjectUser already has. Any particular reason why you're doing this?

Comment (QuanMPhm) [reply src/coldfront_plugin_api/scim_v2/groups.py:139]: @knikolla I didn't realize that. Actually, this makes me want to ask...
Currently I've written the function such that any users who are assigned to allocation are added to the project **with the assumption that they're a user (not a PI, etc) and don't want to be notified**. And as of how I've written it, these will be the assumptions whether the user was never assigned to the project (the if clause evaluates false), or was (if clause is true). 

Do you think that it's fine for the false block to make these assumptions? I'll change the true block to only change status.

File: src/coldfront_plugin_api/scim_v2/groups.py (hunk 8 above)
Comment (QuanMPhm) [reply src/coldfront_plugin_api/scim_v2/groups.py:94]: @knikolla False was just the value I decided to put it. I should have asked you what you'd like it to be. 
Reviewing the Coldfront code, they've set the default `enable_notifications` value to `True`. Should I just leave it like that?

File: src/coldfront_plugin_api/scim_v2/groups.py [hunk 10]
```diff
@@ -80,12 +83,18 @@ def patch(self, request, pk, format=None):
...
+                    )
+                    
                     au = self._set_user_status_on_allocation(
                         allocation, user, "Active"
                     )
                     signals.allocation_activate_user.send(
                         sender=self.__class__, allocation_user_pk=au.pk,
                     )
+
+
```
Comment (knikolla) [inline src/coldfront_plugin_api/scim_v2/groups.py:97]: remove one of these blank lines. two is a bit too much :)

File: src/coldfront_plugin_api/scim_v2/groups.py [hunk 11]
```diff
@@ -117,3 +126,24 @@ def _set_user_status_on_allocation(allocation, user, status):
...
+    
+    @staticmethod
+    def _set_user_status_on_project(project, user, status, role, enable_notifications):
+        pu = ProjectUser.objects.filter(
+            project=project,
+            user=user
+        ).first()
+
+        if pu:
+            pu.status = ProjectUserStatusChoice.objects.get(name=status)
```
Comment (knikolla) [inline src/coldfront_plugin_api/scim_v2/groups.py:138]: is it necessary to `pu.save()` here? I see that in the similar function above I'm calling that method for `au`.

Comment (QuanMPhm) [reply src/coldfront_plugin_api/scim_v2/groups.py:137]: Apparently I don't need to save for `pu`. I don't know why that's the case. Reading the docs doesn't indicate that the save() function saves every object created, not just the object that called it, but I guess we don't have time to look into that?

**Comment by QuanMPhm** [comment]
@knikolla I made the changes you asked for. Though I noticed that the functions we wrote won't check if the usernames provided are valid. Is that going be worth the time addressing?

File: src/coldfront_plugin_api/scim_v2/groups.py (hunk 11 above)
Comment (knikolla) [reply src/coldfront_plugin_api/scim_v2/groups.py:137]: https://docs.djangoproject.com/en/5.0/topics/db/queries/#saving-changes-to-objects indicates that objects aren't saved until `save()` is called. So in this case I would advise erring on the side of caution and calling `save()`.

**Comment by knikolla** [comment]
> @knikolla I made the changes you asked for. Though I noticed that the functions we wrote won't check if the usernames provided are valid. Is that going be worth the time addressing?

We do check that a user with that username exists:

```
try:
    user = utils.get_or_fetch_user(username=submitted_user)
except ObjectDoesNotExist:
    return Response(status=400)
```

and the `_set_user_status_on_project` method takes a user object as an argument, rather than a username.

File: src/coldfront_plugin_api/scim_v2/groups.py [hunk 12]
```diff
@@ -80,12 +83,17 @@ def patch(self, request, pk, format=None):
                     except ObjectDoesNotExist:
                         return Response(status=400)
 
+                    self._set_user_on_project(
```
Comment (knikolla) [inline src/coldfront_plugin_api/scim_v2/groups.py:86]: You've renamed this :)

**Review by knikolla — CHANGES_REQUESTED** [review]

**Comment by QuanMPhm** [comment]
@knikolla Function renamed

File: src/coldfront_plugin_api/scim_v2/groups.py (hunk 8 above)
Comment (knikolla) [reply src/coldfront_plugin_api/scim_v2/groups.py]: Do set this to True. I looked into the production ColdFront and it seems that currently all users on projects have this set to true.

File: src/coldfront_plugin_api/scim_v2/groups.py (hunk 9 above)
Comment (knikolla) [reply src/coldfront_plugin_api/scim_v2/groups.py]: The assumption of the `User` role is correct. As for notifications, I believe it should be set to true.

**Comment by QuanMPhm** [comment]
@knikolla I've made the final changes

File: src/coldfront_plugin_api/scim_v2/groups.py [hunk 13]
```diff
@@ -143,5 +143,6 @@ def _set_user_status_on_project(project, user, status, role, enable_notification
                 role=ProjectUserRoleChoice.objects.get(name=role),
                 enable_notifications = enable_notifications
             )
+            pu.save()
```
Comment (knikolla) [inline src/coldfront_plugin_api/scim_v2/groups.py:146]: This is the wrong place for the save. It should be on update, after line 137, not on create.

**Review by knikolla — APPROVED** [review]
